superset/mcp_service/auth.py

```diff
@@ -119,61 +154,92 @@ def mcp_auth_hook(tool_func: F) -> F:
     """
     Authentication and authorization decorator for MCP tools.
 
-    This is a minimal implementation that:
-    1. Gets the current user
-    2. Sets g.user for Flask context
-
-    TODO (future PR): Add permission checking
-    TODO (future PR): Add JWT scope validation
-    TODO (future PR): Add comprehensive audit logging
-    TODO (future PR): Add rate limiting integration
+    This hook ensures proper Flask app context and user authentication for all
+    MCP tool calls. Deployers can extend this decorator to add:
+
+    - Permission checks: Validate user has specific roles or tool access rights
+    - JWT scope validation: Check OAuth scopes match required tool permissions
+    - Audit logging: Track tool usage, parameters, and results for compliance
+    - Rate limiting: Enforce per-user or per-tool rate limits
+    - Custom middleware: Inject organization-specific security policies
+
+    Default behavior:
+    1. Ensures Flask app context is active (critical for stateless_http mode)
+    2. Authenticates user via get_user_from_request()
+    3. Sets g.user for Superset's security manager integration
+    4. Manages database session lifecycle
+
+    Example extensions:
+        # Add permission check before tool execution:
+        if tool_func.__name__ in ['execute_sql'] and not user.is_admin():
+            raise PermissionError("SQL execution requires admin role")
+
+        # Add audit logging after successful execution:
+        audit_log.info({
+            'user': user.username,
+            'tool': tool_func.__name__,
+            'timestamp': datetime.now(),
+            'result': 'success'
+        })
     """
     import functools
 
     @functools.wraps(tool_func)
     def wrapper(*args: Any, **kwargs: Any) -> Any:
         from superset.extensions import db
+        from superset.mcp_service.flask_singleton import get_flask_app
 
-        # Get user and set Flask context OUTSIDE try block
-        user = get_user_from_request()
+        # Get Flask app instance
+        app = get_flask_app()
 
-        # Force load relationships NOW while session is definitely active
-        _ = user.roles
-        if hasattr(user, "groups"):
-            _ = user.groups
+        # Ensure Flask app context is active
+        # This is critical for stateless_http mode where context isn't preserved
+        with app.app_context():
+            # Get user and set Flask context OUTSIDE try block
+            user = get_user_from_request()
 
-        g.user = user
+            # Force load relationships NOW while session is definitely active
+            _ = user.roles
```

Comment:
I'd be somewhat cautious about using _ as a variable name because it's usually used for the translation function in Superset. Could be confusing.